damus/Views/DMChatView.swift

```diff
@@ -12,21 +12,78 @@ struct DMChatView: View {
+        case message
+    }
+    @FocusState private var focusedField: FocusedField?
+
+    let placeholder = "Send a message"
+    
     var pubkey: String {
         dms.pubkey
     }
-    
+
     var Messages: some View {
         ScrollViewReader { scroller in
             ScrollView {
-                VStack(alignment: .leading) {
-                    ForEach(Array(zip(dms.events, dms.events.indices)), id: \.0.id) { (ev, ind) in
-                        DMView(event: dms.events[ind], damus_state: damus_state)
-                            .contextMenu{MenuItems(event: ev, keypair: damus_state.keypair, target_pubkey: ev.pubkey, bookmarks: damus_state.bookmarks, muted_threads: damus_state.muted_threads)}
-                    }
-                    EndBlock(height: 80)
+                if (dms.events.isEmpty) {
+                    Text("Send a message to start the conversation...", comment: "Text prompt for user to send a message to the other user.")
+                        .lineLimit(nil)
+                        .multilineTextAlignment(.center)
+                        .padding(.horizontal, 40)
+                        .padding(.vertical)
+                        .foregroundColor(.gray)
+                } else {
+                    VStack(alignment: .leading) {
+                        let groups = Dictionary(grouping: dms.events) { event in
+                            Calendar.current.startOfDay(for: Date(timeIntervalSince1970: TimeInterval(event.created_at)))
+                        }.mapValues { events in
+                            var lastEventDate: Date?
+                            var groupCounter = 0
+                            
+                            var eventGroups = [String: [NostrEvent]]()
+                            
+                            for event in events {
+                                let currentEventDate = Date(timeIntervalSince1970: TimeInterval(event.created_at))
+                                defer { lastEventDate = currentEventDate }
+                                
+                                let pubkey = event.pubkey
+                                if let lastDate = lastEventDate,
+                                   currentEventDate.timeIntervalSince(lastDate) <= 120,
+                                   let lastEvent = eventGroups["\(groupCounter)-\(pubkey)"]?.last,
+                                   lastEvent.pubkey == pubkey {
+                                    eventGroups["\(groupCounter)-\(pubkey)"]?.append(event)
+                                } else {
+                                    groupCounter += 1
+                                    eventGroups["\(groupCounter)-\(pubkey)"] = [event]
+                                }
+                            }
+                            
+                            return eventGroups.mapValues { $0.sorted(by: { $0.created_at < $1.created_at }) }
+                        } as [Date: [String: [NostrEvent]]]
```

Comment:
what does this code do ?

Reply:
It separates the messages into groups. Groups based on day and messages sent within 2 minutes of each other (so that it displays those messages together).